ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs

```diff
@@ -56,63 +60,91 @@ import           Ouroboros.Network.Mux.Types
+    :: (ptcl -> Channel m ByteString -> m a)
+    -> MuxApplication ptcl m
+
+  MuxServerApplication
+    -- Server application; similarly to the @'MuxClientApplication'@ but it
+    -- will be run using @'muxResponder'@.
+    :: (ptcl -> Channel m ByteString -> m a)
+    -> MuxApplication ptcl m
+
+  MuxClientAndServerApplication
+    -- Client and server applications.
+    :: (ptcl -> Channel m ByteString -> m a)
+    -> (ptcl -> Channel m ByteString -> m b)
+    -> MuxApplication ptcl m
+
+
+-- |
+-- This type is only necessary to use the @'simpleMuxClient'@ and
+-- @'simpleMuxServer'@ smart constructors.
+data MuxPeer failure m where
+    MuxPeer :: Tracer m (TraceSendRecv ps)
+            -> Codec ps failure m ByteString
+            -> Peer ps pr st m a
+            -> MuxPeer failure m
+
+    MuxPeerPipelined
+            :: Natural
+            -> Tracer m (TraceSendRecv ps)
+            -> Codec ps failure m ByteString
+            -> PeerPipelined ps pr st m a
+            -> MuxPeer failure m
+
 -- |
--- Specification for peers of a protocol.  This type instructs the multiplexing
--- layer to run a given client \/ server peers.
---
-data MuxPeer m where
-  -- |
-  -- A non pipeliend peer together with a codec.
-  --
-  OnlyClient
-    :: forall ps (st :: ps) m a.
-       Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> Peer ps AsClient st m a
-    -> MuxPeer m
-
-  -- |
-  -- A pipelined peer together with a codec.
-  --
-  OnlyPipelinedClient
-    :: forall ps (st :: ps) m a.
-       Natural
-    -> Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> PeerPipelined ps AsClient st m a
-    -> MuxPeer m
-
-  -- |
-  -- Server peer with a codec
-  OnlyServer
-    :: forall ps (st :: ps) m a.
-       Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
-
-  -- |
-  -- Client and server peers with the corresponding codec.
-  --
-  ClientAndServer
-    :: forall ps (st :: ps) m a.
-       Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> Peer ps AsClient st m a
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
-
-  -- |
-  -- Pipelined client and a server with the correspnding codec.
-  --
-  PipelinedClientAndServer
-    :: forall ps (st :: ps) m a.
-       Natural
-    -> Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> PeerPipelined ps AsClient st m a
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
+-- Smart constructor for @'MuxClientApplication'@.  It is a simple client, since
+-- none of the applications requires resource handling to run in the monad @m@.
+-- Each one is simply run either by @'runPeer'@ or @'runPipelinedPeer'@.
+--
+simpleMuxClientApplication
+  :: MonadThrow m
+  => MonadCatch m
+  => MonadAsync m
+  => Exception failure
+  => (ptcl -> MuxPeer failure m)
```

Comment:
Ha, I didn't even know it was allowed to put multiple `=>`s

Reply:
I kinda like it

Reply:
I've learned it a few weeks ago; it's also the only supported syntax in PureScript.